MINOR: Maybe decorate inner topics for SourceNode

guozhangwang opened the pull request:
When creating the source node in TopologyBuilder, we need to decorate its input topics if they are inner (i.e. repartition) topics with the prefix.

Also did some minor cleanup in the printing function for better visualization in debugging.

guozhangwang committed:
Maybe decorate inner topics for SourceNode

guozhangwang commented:
@dguy @mjsax @enothereska for reviews.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:
```diff
@@ -86,16 +86,18 @@ public void process(final K key, final V value) {
-            sb.append("stateStores [");
-            for (String store : (Set<String>) stateStores) {
-                sb.append(store + ",");
+            sb.append("\n\t\t\t\t\tstates:\t\t[");
```
Looks like a lot of taps...  Should we have a "indent level" variable that we track and increase the indent level while calling nested `toString()` to simplify this?

guozhangwang replied:
I thought about it before, but it is a bit tricky to do since it could be dynamic based on which class the `toString` function is triggered first.  Let me think about it more and see if I can come with a clean solution.

mjsax replied:
Not sure what you mean by tricky. You just hand in the current indention string as parameter and each time you call a nested `toString(String indent)` you add one tab.

guozhangwang replied:
`toString(String)` is not overridable Object function. Anyways, I did a bit hacky way to have each class embedding a private `toString(String)` function. See if you like it.

mjsax reviewed (commented):
Overall LGTM. One suggestion to improve indention tracking.

guozhangwang commented:
test this.

guozhangwang commented:
test this

guozhangwang commented:
please test this

guozhangwang commented:
Example output with two sub-topologies:

```
ProcessorTopology:
	KSTREAM-SOURCE-0000000000:
		topics:		[topic1]
		children:	[KSTREAM-MAP-0000000001]
	KSTREAM-MAP-0000000001:
		children:	[KSTREAM-FILTER-0000000004, KSTREAM-FILTER-0000000007]
	KSTREAM-FILTER-0000000004:
		children:	[KSTREAM-SINK-0000000003]
	KSTREAM-SINK-0000000003:
		topic:		X-Counts-repartition
	KSTREAM-FILTER-0000000007:
		children:	[KSTREAM-SINK-0000000006]
	KSTREAM-SINK-0000000006:
		topic:		X-KSTREAM-MAP-0000000001-repartition

ProcessorTopology:
	KSTREAM-SOURCE-0000000008:
		topics:		[X-KSTREAM-MAP-0000000001-repartition]
		children:	[KSTREAM-LEFTJOIN-0000000009]
	KSTREAM-LEFTJOIN-0000000009:
		states:		[Counts]
	KSTREAM-SOURCE-0000000005:
		topics:		[X-Counts-repartition]
		children:	[KSTREAM-AGGREGATE-0000000002]
	KSTREAM-AGGREGATE-0000000002:
		states:		[Counts]

```

guozhangwang committed:
add indent to toString

dguy commented:
Test failures look like they are legit. Same in each build:
org.apache.kafka.streams.processor.internals.StreamThreadTest.shouldCloseActiveTasksThatAreAssignedToThisStreamThreadButAssignmentHasChangedBeforeCreatingNewTasks

dguy commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -389,10 +389,21 @@ public void run() {
+    }
+
+    /**
+     * Produces a string representation contain useful information about Kafka Streams
```
"representation contain[ing]..."

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -188,7 +188,8 @@ private SourceNodeFactory(String name, String[] topics, Pattern pattern, Deseria
         @Override
         public ProcessorNode build() {
-            return new SourceNode(name, nodeToSourceTopics.get(name), keyDeserializer, valDeserializer);
+            String[] sourceTopics = nodeToSourceTopics.get(name);
```
I know this is a bit opinionated, but ... I think we should make an effort to make all locals `final` where possible. It is just a few extra keystrokes (that intellij can do for you!), and it helps to eliminate a class of bugs.

guozhangwang replied:
Totally!

mjsax replied:
I guess doing this incrementally will not work. What about one (or multiple) PR that add `final` wherever possible over all files (maybe package by package to split it up in multiple PRs)? It's dumb work, but I would volunteer...

If we do in incrementally, it also distracts from actual code changes for a PR and add noise that makes reviewing harder.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -161,25 +161,33 @@ public StateStore getStore(final String name) {
+    }
+
+    /**
+     * Produces a string representation contain useful information about a StreamTask starting with the given indent.
```
as above "contain[ing]"

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java:
```diff
@@ -86,40 +86,49 @@ public SinkNode sink(String topic) {
 
     /**
-     * Produces a string representation contain useful information this topology.
+     * Produces a string representation contain useful information this topology starting with the given indent.
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorTopology.java:
```diff
@@ -86,40 +86,49 @@ public SinkNode sink(String topic) {
+    }
+
+    /**
+     * Produces a string representation contain useful information this topology.
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1031,25 +1031,34 @@ public void apply(final AbstractTask task) {
+    }
+
+    /**
+     * Produces a string representation contain useful information about a StreamThread, starting with the given indent.
```
As above

dguy reviewed (commented):
Left a few minor comments.
I have one general comment that is not specific to this PR:
I'd really like to see us using `final` as much as possible. It should be used for all method params. It should also be used for all fields and locals where possible. It helps to eliminate bugs and shows intent.

guozhangwang committed:
Fix the build() function when it is called before topics metadata updated for Pattern subscription

guozhangwang commented:
Addressed commends. Also fixed the issue when `build()` function is called before the topics metadata has been refreshed for pattern subscriptions (this should only happen when users explicitly call `builder.build()`, during debug / testing).

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into KMinor-source-topic-fix

guozhangwang committed:
checkstyle fix

enothereska commented:
LGTM thanks

junrao reviewed (commented):
Thanks for the patch. LGTM

guozhangwang commented:
Merged to trunk.
